### PR TITLE
More packaging fixes

### PR DESCRIPTION
**Pull request by ssbarnea**
- Use modern build packaging
- Correct errors from Manifest.in which caused accidental inclusion
  of .mypy_cache folders from various locations in source tree
- Avoid tox bug with editale when setup.py is missing
- Final archive should now be less than 300kb

Related: https://github.com/tox-dev/tox/issues/2197

### CODE REVIEW

File: MANIFEST.in [hunk 1]
```diff
@@ -1,24 +1,31 @@
...
-include *.py
-include *.txt
-include *.yaml
-include *.yml
+prune zuul.d
+prune .github
+exclude .git_archival.txt
+exclude playbooks
+exclude snap
+include src/molecule/test/functional/.ansible-lint
```
Comment (webknjaz) [inline MANIFEST.in:7]: It's pointless to enumerate each file since they are already autoincluded by setuptools-scm.

File: MANIFEST.in [hunk 2]
```diff
@@ -1,24 +1,31 @@
 prune asset
```
Comment (webknjaz) [inline MANIFEST.in:1]: Looks like this line is the only entry that actually makes sense to be in this file.

File: MANIFEST.in [hunk 3]
```diff
@@ -31,7 +38,4 @@ recursive-include snap *.yaml
 recursive-include tools *.sh
 recursive-include src/molecule/test/scenarios extra-host
 recursive-include src/molecule/test/scenarios hosts
-recursive-exclude .github *.*
-recursive-exclude .mypy_cache *.*
-recursive-exclude docs/docstree/html *.*
-recursive-exclude zuul.d *.*
+recursive-exclude docs/docstree/html/_static *.*
```
Comment (webknjaz) [inline MANIFEST.in:41]: Is this the built content?

File: tox.ini [hunk 4]
```diff
@@ -155,20 +158,18 @@ usedevelop = false
 # don't install molecule itself in this env
 skip_install = true
 deps =
-    check-manifest
+    check-manifest >= 0.47
     collective.checkdocs >= 0.2
-    pep517 >= 0.8.2
+    build >= 0.7.0
```
Comment (webknjaz) [inline tox.ini:163]: I'd rather keep this capped to the next minor.

**Review by webknjaz — CHANGES_REQUESTED** [review]
There's a lot of unnecessary lines in the manifest. Drop them.

File: MANIFEST.in [hunk 5]
```diff
@@ -1,24 +1,31 @@
 prune asset
-include *.md
-include *.py
-include *.txt
-include *.yaml
-include *.yml
+prune zuul.d
```
[the remark this answers is not in the document]
Comment (ssbarnea) [reply MANIFEST.in:2]: I do not see any use to include CI config in sdist, same with .github folder. On the other hand I kept test related files  so you can still run tests from source archive.

File: MANIFEST.in (hunk 2 above)
Comment (ssbarnea) [reply MANIFEST.in:1]: That is far from true, i tried this myself but the result caused problems in multiple places, one of them was the check-manifest. The other was that extra built docs were included and that desired json/yaml were not included from inside src/molecule. On the other hand if you put generic recursive includes for those two extensions you may endup getting stuff like `src/molecule/test/.mypy_cache/.../foo.json` which are excluded by default.

I know that testing behavior of Manifest.in is quite tricky as it is highly dependent on what you may have on disk, like temp files, stuff that are not VCS tracked. 

`check-manifest` command is very useful for debugging as it reports both kind of unexpected exclusions or inclusions but still fails to give correct recommendations or to spot the bug I mentioned.

File: MANIFEST.in (hunk 3 above)
Comment (ssbarnea) [reply MANIFEST.in:41]: yeah

File: MANIFEST.in (hunk 5 above)
[the remark this answers is not in the document]
Comment (webknjaz) [reply MANIFEST.in:5]: The whole point of the sdist is to provide the sources for stuff, as much as possible. It's right there in the name. The users needing a small thing will use the wheel anyway (and it's configured separately). Many things use sdists as the source of truth, in particular — the downstream packagers. So don't exclude things without a really compelling reason.

**Comment by ssbarnea** [comment]
@webknjaz If you can find some reliable source like official packaging docs claiming that CI/CD config should be included in sdist I will do that. AFAIK, the size is still important and we should not include stuff in bulk. I am not saying I need to remove every file but at least those folders that can influence the final size in a significant way.

I really doubt any downstream maintainer relies on sdist archives instead of having a fork of the upstream repository. It would be insane to maintain an active package using only sdists.

In fact most people strip even the tests from sdist, but we do not.

File: MANIFEST.in (hunk 2 above)
Comment (webknjaz) [reply MANIFEST.in:2]: Oh, so your problem was because of the recursive inclusion and not a bug in `setuptools-scm`. That makes sense, then.

**Comment by webknjaz** [comment]
> @webknjaz If you can find some reliable source like official packaging docs claiming that CI/CD config should be included in sdist I will do that. AFAIK, the size is still important and we should not include stuff in bulk. I am not saying I need to remove every file but at least those folders that can influence the final size in a significant way.

I'm a reliable source. A lot of things are still tribal knowledge, unfortunately. But I prefer to follow what the setuptools maintainer does for reducing the maintenance burden of looking after hundreds of projects on PyPI.
I agree that some extra excludes can be useful but if it doesn't cause problems reported by the actual users, I'd rather not give up the ease of maintaining this packaging setup.

> I really doubt any downstream maintainer relies on sdist archives instead of having a fork of the upstream repository. It would be insane to maintain an active package using only sdists.

Well, I was speaking based on my experience and prior research. I can only speak about Gentoo and Fedora but they do exactly this — they point at sdists in the corresponding package specs of the respective ecosystems: https://github.com/gentoo/gentoo/blob/39729ace/app-admin/ansible-lint/ansible-lint-5.2.1.ebuild#L12 + https://src.fedoraproject.org/rpms/python-molecule/blob/c93b515b/f/python-molecule.spec#_20.
I attended PackagingCon three weeks ago and there were packagers from a lot of ecosystems, including distros and they all were in agreement on this.

> In fact most people strip even the tests from sdist, but we do not.

And when they do, the downstream maintainers get especially unhappy about this. They come to the upstream projects asking to fix this problem. This was specifically emphasized in the talk of the Fedora's Python packagers, brought up and confirmed in the conversations on matrix, and I've heard this in the past from downstream maintainers in other places on the internet.

**Comment by webknjaz** [comment]
Duplicate of #3324

**Pull request closed without merging**